synapse/util/file_consumer.py

```diff
@@ -0,0 +1,158 @@
+    """A consumer that writes to a file like object. Supports both push
+    and pull producers
+
+    Args:
+        file_obj (file): The file like object to write to. Closed when
+            finished.
+    """
+
+    # For PushProducers pause if we have this many unwritten slices
+    _PAUSE_ON_QUEUE_SIZE = 5
+    # And resume once the size of the queue is less than this
+    _RESUME_ON_QUEUE_SIZE = 2
+
+    def __init__(self, file_obj):
+        self.file_obj = file_obj
+
+        # Producer we're registered with
+        self.producer = None
+
+        # True if PushProducer, false if PullProducer
+        self.streaming = False
+
+        # Queue of slices of bytes to be written. When producer calls
+        # unregister a final None is sent.
+        self.bytes_queue = Queue.Queue()
+
+        # Deferred that is resolved when finished writing
+        self.finished_deferred = None
+
+        # If the _writer thread throws an exception it gets stored here.
+        self._write_exception = None
+
+        # A deferred that gets resolved when the bytes_queue gets empty.
+        # Mainly used for tests.
+        self._notify_empty_deferred = None
+
+    def registerProducer(self, producer, streaming):
+        """Part of IProducer interface
+
+        Args:
+            producer (IProducer)
+            streaming (bool): True if push based producer, False if pull
+                based.
+        """
+        self.producer = producer
+        self.streaming = streaming
+        self.finished_deferred = threads.deferToThread(self._writer)
+        if not streaming:
+            self.producer.resumeProducing()
+
+        self.paused_producer = False
+
+    def unregisterProducer(self):
+        """Part of IProducer interface
+        """
+        self.producer = None
+        if not self.finished_deferred.called:
+            self.bytes_queue.put_nowait(None)
+
+    def write(self, bytes):
+        """Part of IProducer interface
+        """
+        if self._write_exception:
+            raise self._write_exception
+
+        if self.finished_deferred.called:
+            raise Exception("consumer has closed")
+
+        self.bytes_queue.put_nowait(bytes)
+
+        # If this is a pushed based consumer and the queue is getting behind
+        # then we pause the producer.
+        if self.streaming and self.bytes_queue.qsize() >= self._PAUSE_ON_QUEUE_SIZE:
+            self.paused_producer = True
+            self.producer.pauseProducing()
+
+    def _writer(self):
+        """This is run in a background thread to write to the file.
+        """
+        try:
+            while self.producer or not self.bytes_queue.empty():
+                # If we've paused the producer check if we should resume the
+                # producer.
+                if self.producer and self.paused_producer:
+                    if self.bytes_queue.qsize() <= self._RESUME_ON_QUEUE_SIZE:
+                        reactor.callFromThread(self._resume_paused_producer)
```

Comment:
does this block until the callee finishes, or not? if not we could end up resuming multiple times - is that a problem?

Reply:
I'm fairly sure it doesn't, and its not a problem as `_resume_paused_producer` checks `self.paused_producer` again anyway.